### ISSUE TITLE
Custom templates for notification emails are broken

### ISSUE THREAD

**Issue by stevekm** [nextflow-io/nextflow#4578]
## Bug report 

According to the docs, you can define a custom email template like this

https://www.nextflow.io/docs/latest/config.html#scope-notification

```groovy
notification {
    enabled = true
    to = "${params.emailTo}"
    from = "${params.emailFrom}"
    template = "assets/mail/notification.html"
}
```

However it does not work.

Using the default notification template file from here; https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/resources/nextflow/mail/notification.html copied as the file `assets/mail/notification.html`. It breaks with the error:

```
Dec-12 16:15:46.131 [main] WARN  nextflow.script.WorkflowMetadata - Failed to deliver notification email -- See the log file for details
java.lang.NullPointerException: Cannot invoke "java.util.Map.size()" because "m" is null
        at java.base/java.util.HashMap.putMapEntries(HashMap.java:503)
        at java.base/java.util.HashMap.putAll(HashMap.java:791)
        at nextflow.script.WorkflowNotifier.loadMailTemplate0(WorkflowNotifier.groovy:222)
        at nextflow.script.WorkflowNotifier.loadMailTemplate(WorkflowNotifier.groovy:183)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1258)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
        at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1036)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1019)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:97)
        at nextflow.script.WorkflowNotifier$_createMail_closure1.doCall(WorkflowNotifier.groovy:116)
        at nextflow.script.WorkflowNotifier$_createMail_closure1.call(WorkflowNotifier.groovy)
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2357)
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2342)
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2383)
        at nextflow.script.WorkflowNotifier.createMail(WorkflowNotifier.groovy:115)
        at nextflow.script.WorkflowNotifier.sendNotification(WorkflowNotifier.groovy:69)
        at nextflow.script.WorkflowMetadata.safeMailNotification(WorkflowMetadata.groovy:460)
        at nextflow.script.WorkflowMetadata.invokeOnComplete(WorkflowMetadata.groovy:396)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1258)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
        at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1036)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1019)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:97)
        at nextflow.script.WorkflowMetadata$_closure3.doCall(WorkflowMetadata.groovy:251)
        at nextflow.script.WorkflowMetadata$_closure3.call(WorkflowMetadata.groovy)
        at groovy.lang.Closure.run(Closure.java:493)
        at nextflow.Session.shutdown0(Session.groovy:694)
        at nextflow.Session.destroy(Session.groovy:644)
        at nextflow.script.ScriptRunner.shutdown(ScriptRunner.groovy:239)
        at nextflow.script.ScriptRunner.execute(ScriptRunner.groovy:134)
        at nextflow.cli.CmdRun.run(CmdRun.groovy:368)
        at nextflow.cli.Launcher.run(Launcher.groovy:494)
        at nextflow.cli.Launcher.main(Launcher.groovy:653)

```

It appears to be breaking when it tries to call this function [`loadMailTemplate0`](https://github.com/nextflow-io/nextflow/blob/832bff241f64cc3ce9b4cdc794a58d1664c0933c/modules/nextflow/src/main/groovy/nextflow/script/WorkflowNotifier.groovy#L219-L226)

which is called [here by `loadMailTemplate`](https://github.com/nextflow-io/nextflow/blob/832bff241f64cc3ce9b4cdc794a58d1664c0933c/modules/nextflow/src/main/groovy/nextflow/script/WorkflowNotifier.groovy#L183)

which appears to be called [here while loading user mail templates](https://github.com/nextflow-io/nextflow/blob/832bff241f64cc3ce9b4cdc794a58d1664c0933c/modules/nextflow/src/main/groovy/nextflow/script/WorkflowNotifier.groovy#L109-L122)

it sounds like an issue with this `binding` object which is getting passed about; I noticed that there is a listing for "binding" in the docs

> notification.binding
> 
>  An associative array modelling the variables in the template file.

I tried to update my config like this

```groovy
notification {
    enabled = true
    to = "${params.emailTo}"
    from = "${params.emailFrom}"
    template = "assets/mail/notification.html"
    binding = [ "workflow": workflow ]
}
```

However this does not work for a variety of reasons such as 

```
  class java.util.LinkedHashMap cannot be cast to class groovy.lang.Binding (java.util.LinkedHashMap is in module java.base of loader 'bootstrap'; groovy.lang.Binding is in unnamed module of loader 'app')
```

This also does not work

```
binding = new Binding("workflow": workflow)
```

Its worth noting that it seems like the default email template relies on extensive usage of the `workflow` object, but this object is not actually usable from within the Nextflow config file (?). So even if this binding did work, it seems like you would not actually be able to anything from `workflow` in your custom email notification from the Nextflow config. Unless there is some other method you are supposed to use to access the `workflow` object from the config in order to use it with custom emails?

Let me know if there are examples of how to use the custom email template that work, so far I have been searching and have not been able to find any. It seems like some of the nf-core pipelines had their own alternative implementation of custom emails that circumvented these issues.

### Environment 

* Nextflow version:  23.04.1 build 5866
* Java version: openjdk version "20-internal" 2023-03-21
* Operating system: Ubuntu 20.04
* Bash version: 5.0.17

**Comment by stevekm** [comment]
also same error with Nextflow version `23.10.0 build 5889`

**Comment by pditommaso** [comment]
OK, solved by ccf4f59e. Thanks for reporting it

**Comment by stevekm** [comment]
Thanks! would it be possible to get a minimal example of the syntax for this? Is it supposed to be like this?

```groovy
notification {
    template = "assets/mail/notification.html"
    attributes = [ workflow: workflow ]
}
```

also is the `workflow` object supposed to be usable here? I have had difficulties in the past trying to use it inside various config scopes.
